### PR TITLE
Refactor PyTorch model reader.

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
Refactor includes:
1. remove duplication of framework_attr and node.kwargs in FunctionalNodes
2. reorder graph_builders.py
3. fix substitutions following the above changes

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 1]
```diff
@@ -53,23 +54,22 @@ def get_attributes_from_weights(node: BaseNode) -> Dict:
                         GAMMA: np.ones(node.weights[1].shape),
                         BETA: np.zeros(node.weights[1].shape)}
 
-        has_weight = WEIGHT not in node.framework_attr
-        has_bias = BIAS not in node.framework_attr
+        # Check if weight and/or bias were not given.
+        if KERNEL in node.tensor_input_allocs:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:58]: is node a BN node? if so, why would it have kernel and bias?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 2]
```diff
@@ -53,23 +54,22 @@ def get_attributes_from_weights(node: BaseNode) -> Dict:
...
+            weights_dict[GAMMA] = node.weights[node.tensor_input_allocs.index(KERNEL)]
+        elif KERNEL not in node.op_call_kwargs:
+            weights_dict[GAMMA] = node.weights[3]
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:61]: same question regarding the arbitrary constant "3"
(also, consider making these constant variables)

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py [hunk 3]
```diff
@@ -37,7 +38,7 @@ def __init__(self):
         super().__init__(matcher_instance=bn_node)
 
     @staticmethod
-    def get_attributes_from_weights(node: BaseNode) -> Dict:
+    def get_attributes_from_weights(node: FunctionalNode) -> Dict:
         """
         convert functional batch_norm positional weights to BatchNorm2d weights
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_batch_norm.py:43]: I dont understand why would the BN weights be positional and not regular parameters of the layer (maybe talk about this offline)

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_layer_norm.py [hunk 4]
```diff
@@ -38,7 +39,7 @@ def __init__(self):
         super().__init__(matcher_instance=ln_node)
 
     @staticmethod
-    def get_attributes_from_weights(node: BaseNode, normalized_shape: [Tuple, List, int]) -> Dict:
+    def get_attributes_from_weights(node: FunctionalNode, normalized_shape: [Tuple, List, int]) -> Dict:
         """
         Parse layer_norm(input, normalized_shape, weight=None, bias=None)
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_layer_norm.py:44]: need to elaborate more (this doesn't explain what the function is doing)

File: tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py [hunk 5]
```diff
@@ -81,7 +85,11 @@ def compare(self, quantized_models, float_model, input_x=None, quantization_info
 
         is_bn_in_model = nn.BatchNorm2d in [type(module) for name, module in quant_model.named_modules()]
         self.unit_test.assertTrue(self.fold_applied is not is_bn_in_model)
-        self.unit_test.assertTrue(np.isclose(out_quant, out_float, atol=1e-5, rtol=1e-4).all())
+
+        # TODO: remove this and set atol=1e-5 for all tests after fixing the issue with ConvTranspose2d.
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py:89]: attach a link to the issue on github in the comment

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/const_holder_conv.py [hunk 6]
```diff
@@ -60,9 +61,11 @@ def substitute(self,
         # Create new node of layer convolution
         if 1 not in func_node.weights:
             Logger.critical(f'Weight input missing for node {func_node.name}.')  # pragma: no cover
-        weight = func_node.weights[1]
-        bias = func_node.weights.get(2)
-        framework_attr = func_node.framework_attr
+        weight_index = func_node.tensor_input_allocs.index(KERNEL) if KERNEL in func_node.tensor_input_allocs else 1
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/const_holder_conv.py:64]: default in the framework if weight and bias are in the args.

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/const_holder_conv.py [hunk 7]
```diff
@@ -60,9 +61,11 @@ def substitute(self,
...
+        bias_index = func_node.tensor_input_allocs.index(BIAS) if BIAS in func_node.tensor_input_allocs else 2
+        weight = func_node.weights[weight_index]
+        bias = func_node.weights.get(bias_index)
+        framework_attr = func_node.op_call_kwargs
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/const_holder_conv.py:68]: Converting functional conv to conv module, so moving the kwargs to framework_attr, as the convention in the model reader.

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 8]
```diff
@@ -28,29 +30,130 @@
...
+
+
+def is_instance_first_arg(n: Node, expected_type: Any) -> bool:
+    """
+    Check whether first argument of the node is the expected type
+    Args:
+        n: fx node.
+        expected_type: Expected 1st argument type.
+
+    Returns:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:57]: Return

File: model_compression_toolkit/core/pytorch/reader/graph_builders.py [hunk 9]
```diff
@@ -28,29 +30,130 @@
...
+        Dictionary containing module parameters and buffers by name.
+    """
+
+    named_parameters = {name: to_numpy(parameter) for name, parameter in module.named_parameters()}
+    named_buffers = {name: to_numpy(buffer) for name, buffer in module.named_buffers()}
+
+    return {**named_parameters, **named_buffers}
+
+
+def is_instance_first_arg(n: Node, expected_type: Any) -> bool:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/reader/graph_builders.py:50]: expected_type: Type?

**Review by ofirgo — APPROVED** [review]
